src/Cardano/Wallet/AddressDiscovery.hs

```diff
@@ -0,0 +1,229 @@
+    } deriving (Generic, Show, Eq)
+
+instance NFData AddressPool
+
+-- | Get all addresses in the pool, sorted from the first address discovered,
+-- up until the next one.
+--
+-- In practice, we always have:
+--
+-- > mkAddressPool key g cc (addresses pool) == pool
+addresses :: AddressPool -> [Address]
+addresses = map fst . sortOn snd . Map.toList . indexedAddresses
+
+-- | Create a new Address pool from a list of addresses. Note that, the list is
+-- expected to be ordered in sequence (first indexes, first in the list).
+--
+-- The pool will grow from the start if less than @g :: AddressPoolGap@ are
+-- given, such that, there are always @g@ undiscovered addresses in the pool.
+mkAddressPool
+    :: Key 'AccountK XPub
+    -> AddressPoolGap
+    -> ChangeChain
+    -> [Address]
+    -> AddressPool
+mkAddressPool key g cc addrs = AddressPool
+    { accountPubKey = key
+    , gap = g
+    , changeChain = cc
+    , indexedAddresses = nextAddresses key g cc minBound <>
+        Map.fromList (zip addrs [minBound..maxBound])
+    }
+
+
+-- | Lookup an address in the pool. When we find an address in a pool, the pool
+-- may be amended if the address was discovered near the edge. It is also
+-- possible that the pool is not amended at all - this happens in the case that
+-- an address is discovered 'far' from the edge.
+lookupAddress
+    :: Address
+    -> AddressPool
+    -> (Maybe (Index 'Soft 'AddressK), AddressPool)
+lookupAddress !target !pool =
+    case Map.lookup target (indexedAddresses pool) of
+        Just ix ->
+            (Just ix, extendAddressPool ix pool)
+        Nothing ->
+            (Nothing, pool)
+
+-- | If an address is discovered near the edge, we extend the address sequence,
+-- otherwise we return the pool untouched.
+extendAddressPool
+    :: Index 'Soft 'AddressK
+    -> AddressPool
+    -> AddressPool
+extendAddressPool !ix !pool
+    | isOnEdge  = pool { indexedAddresses = indexedAddresses pool <> next }
+    | otherwise = pool
+  where
+    edge = Map.size (indexedAddresses pool)
+    isOnEdge = edge - fromEnum ix <= fromEnum (gap pool)
+    next = if ix == maxBound then mempty else nextAddresses
+        (accountPubKey pool)
+        (gap pool)
+        (changeChain pool)
+        (succ ix)
+
+-- | Compute the pool extension from a starting index
+nextAddresses
+    :: Key 'AccountK XPub
+    -> AddressPoolGap
+    -> ChangeChain
+    -> Index 'Soft 'AddressK
+    -> Map Address (Index 'Soft 'AddressK)
+nextAddresses !key (AddressPoolGap !g) !cc !fromIx =
+    invariant
+        (toIx >= fromIx)
+        "nextAddresses: toIx should be greater than fromIx"
+        safeNextAddresses
+  where
+    safeNextAddresses = [fromIx .. min maxBound toIx]
+        & map (\ix -> (newAddress ix, ix))
+        & Map.fromList
+    toIx = toEnum $ fromEnum fromIx + fromEnum g - 1
+    newAddress = keyToAddress . deriveAddressPublicKey key cc
+
+-- | Fails hard if an invariant does not hold, or proceed.
+invariant
```

Comment:
maybe it is good idea to extract `invariant` function in helper file or use the one already defined in `src/Cardano/Wallet.hs` : 
```
173:invariant
174-    :: String -- ^ A title / message to throw in case of violation
175-    -> a
176-    -> (a -> Bool)
177-    -> a
178:invariant msg a predicate
```

Reply:
Whoops. My bad. I intended to use the one from `src/Cardano/Wallet.hs` indeed, though I hadn't rebase yet at this moment and just copy/pasted it quickly. Will remove :+1: